src/origin-experiments.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {bytesToString, stringToBytes} from './utils/bytes';
+import {getSourceOrigin, parseUrl} from './url';
+import {parseJson} from './json';
+
+/**
+ * Generates, signs and verifies origin experiments.
+ */
+export class OriginExperiments {
+  /**
+   * @param {!./service/crypto-impl.Crypto} crypto
+   */
+  constructor(crypto) {
+    /** @private {!./service/crypto-impl.Crypto} */
+    this.crypto_ = crypto;
+  }
+
+  /**
+   * Generates an RSA public/private key pair for signing and verifying.
+   * @return {!Promise}
+   */
+  generateKeys() {
+    const generationAlgo = Object.assign({
+      modulusLength: 2048,
+      publicExponent: Uint8Array.of(1, 0, 1),
+    }, this.crypto_.pkcsAlgo);
+    return this.crypto_.subtle.generateKey(
+        /** @type {{name: string}} */ (generationAlgo),
+        /* extractable */ true,
+        /* keyUsages */ ['sign', 'verify']);
+  }
+
+  /**
+   * Generates an origin experiment token given a config json.
+   * @param {number} version
+   * @param {!JsonObject} json
+   * @param {!webCrypto.CryptoKey} privateKey
+   * @return {!Promise<string>}
+   */
+  generateToken(version, json, privateKey) {
+    const config = stringToBytes(JSON.stringify(json));
+    const data = this.prepend_(version, config);
+    return this.sign_(data, privateKey).then(signature => {
+      return this.append_(data, new Uint8Array(signature));
+    });
+  }
+
+  /**
+   * Verifies an origin experiment token given a public key.
+   * @param {string} token
+   * @param {!Location} location
+   * @param {!webCrypto.CryptoKey} publicKey
+   * @return {!Promise<string>} If token is valid, resolves with the
+   *     experiment ID. Otherwise, rejects with validation error.
+   */
+  verifyToken(token, location, publicKey) {
+    return new Promise((resolve, reject) => {
+      let i = 0;
+      const bytes = stringToBytes(atob(token));
+
+      // Parse version.
+      const version = bytes[i];
+      if (version !== 0) {
+        reject(new Error(`Unrecognized token version: ${version}`));
```

Comment:
You don't need to reject, especially since it doesn't stop the rest of the function from happening. Throw errors.